0012-unify-pulse-commands-and-instructions.md

```diff
@@ -0,0 +1,334 @@
+
+Communication of changes to users will be performed with a reno entry for the changes and deprecation warnings for deprecated components.
+
+## Detailed Design
+
+### Classes
+#### Pulse
+
+```python
+class Pulse(ABC):
+
+  @property
+  @abstractmethod
+  def duration(self) -> int:
+    ...
+
+  @property
+  @abstractmethod
+  def name(self) -> str:
+    ...
+
+  def __call__(self, channel: PulseChannel) -> Play:
+    return Play(self, channel)
+```
+
+#### Instruction
+```python
+class Instruction(ABC):
+  def __init__(self, *operands) -> Instruction:
+    ...
+
+  @property
+  @abstractmethod
+  def duration(self) -> int:
+    ...
+
+  @property
+  def operands(self) -> Tuple[Any, ...]:
+    return operands
+
+  @property
+  @abstractmethod
+  def channels -> Tuple[Channel, ...]:
+    ...
+```
+
+#### SamplePulse
+```python
+class SamplePulse(Pulse):
+  # otherwise unchanged
+  ...
+```
+
+#### `ParametricPulse`
+```python
+class ParametricPulse(Pulse):
+  # otherwise unchanged
+  ...
+```
+
+#### Play
+```python
+class Play(Instruction):
+  def __init__(self, pulse: Pulse, channel: PulseChannel) -> Play:
+    super().__init__(pulse, channel)
+
+  @property
+  def duration(self) -> int:
+    return pulse.duration
+
+  @property
+  def channels:
+    return (channel,)
+
+  @property
+  def pulse(self) -> Pulse:
+    return pulse
+
+  @property
+  def pulse_name(self) -> str:
+    return pulse.name
+```
+
+#### ShiftPhase
+```python
+class ShiftPhase(Instruction):
+  def __init__(self, phase: float, channel: PulseChannel) -> ShiftPhase:
+    super().__init__(phase, channel)
+
+  @property
+  def duration(self) -> int:
+    return 0
+
+  @property
+  def channels:
+    return (channel,)
+
+  @property
+  def phase(self) -> float:
+    return phase
+```
+
+#### SetPhase
+```python
+class SetPhase(Instruction):
+  def __init__(self, phase: float, channel: PulseChannel) -> SetPhase:
+    super().__init__(phase, channel)
+
+  @property
+  def duration(self) -> int:
+    return 0
+
+  @property
+  def channels:
+    return (channel,)
+
+  @property
+  def phase(self) -> float:
+    return phase
+```
+
+#### SetFrequency
+```python
+class SetFrequency(Instruction):
+  def __init__(self, frequency: float, channel: PulseChannel) -> SetFrequency:
+    super().__init__(frequency, channel)
+
+  @property
+  def duration(self) -> int:
+    return 0
+
+  @property
+  def channels:
+    return (channel,)
+
+  @property
+  def frequency(self) -> float:
+    return frequency
+```
+
+#### ShiftFrequency
+```python
+class ShiftFrequency(Instruction):
+  def __init__(self, frequency: float, channel: PulseChannel) -> ShiftFrequency:
+    super().__init__(frequency, channel)
+
+  @property
+  def duration(self) -> int:
+    return 0
+
+  @property
+  def channels:
+    return (channel,)
+
+  @property
+  def frequency(self) -> float:
```

Comment:
Maybe the ShiftFrequency/Phase Instructions should have frequency/phase_shift properties instead..?

Reply:
Did you mean `frequency_shift/phase_shift`. If so I'm not opposed, thoughts @lcapelluto?

Reply:
I don't have any strong opinion. it would be a positional argument anyway, right? no need to specify the full name when instantiating

Reply:
True - the ctor argument should also be renamed - I was thinking about the possible confusion/quiet bugs arising from the Frequency and ShiftFrequency instructions having equivalent property names - `frequency` (and same for Set/ShiftPhase)

Perhaps there is a benefit to having them be equivalent  - it could be argued they have the same units so a similar name is warranted... up to you